tsdb/index.go

```diff
@@ -1400,20 +1475,35 @@ func (is IndexSet) measurementNamesByExpr(auth query.FineAuthorizer, expr influx
 			return is.measurementNamesByTagFilter(auth, e.Op, tag.Val, value, regex)
 
 		case influxql.OR, influxql.AND:
-			lhs, err := is.measurementNamesByExpr(auth, e.LHS)
+
+			exprToSlice := func(e influxql.Expr) (slice [][]byte, iterator *fileMeasurementSliceIterator, err error) {
+				mi, err := is.measurementNamesByExpr(auth, e)
+				if err != nil {
+					return nil, nil, err
+				}
+				slice, iterator, err = measurementIteratorToSliceIterator(mi)
+				if err != nil {
+					mi.Close()
+					return nil, nil, err
+				}
+				return
+			}
+
+			ls, lsfi, err := exprToSlice(e.LHS)
 			if err != nil {
 				return nil, err
 			}
-
-			rhs, err := is.measurementNamesByExpr(auth, e.RHS)
+			rs, rsfi, err := exprToSlice(e.RHS)
 			if err != nil {
+				lsfi.Close()
 				return nil, err
 			}
 
+			mis := MeasurementIterators{lsfi, rsfi}
 			if e.Op == influxql.OR {
-				return bytesutil.Union(lhs, rhs), nil
+				return newFileMeasurementSliceIterator(bytesutil.Union(ls, rs), mis), nil
```

Comment:
Similarly it might be clearer if this read like:

```
newFileMeasurementSliceIterator(bytesutil.Union(lsfi.UnsafeToSlice(), rsfi.UnsafeToSlice), MeasurementIterators{lsfi, rsfi})
```

Also would it be clearer to name `MeasurementIterators` as `MeasurementIteratorClosers` ?